components/engine/src/lib.rs

```diff
@@ -288,18 +290,21 @@ impl KeyBuilder {
     }
 
     pub fn as_slice(&self) -> &[u8] {
-        unsafe { slice::from_raw_parts(self.as_ptr(), self.len()) }
+        &self.buf.as_slice()[self.start..]
     }
 
-    pub fn build(self) -> Vec<u8> {
+    pub fn build(mut self) -> Vec<u8> {
         if self.start == 0 {
             self.buf
         } else {
-            let mut res = Vec::with_capacity(self.len());
             unsafe {
-                res.extend_from_slice(slice::from_raw_parts(self.as_ptr(), self.len()));
+                ptr::copy(
+                    self.buf.as_ptr().add(self.start),
+                    self.buf.as_mut_ptr(),
+                    self.buf.len(),
+                );
             }
-            res
+            self.buf
```

Comment:
you may need to modify the vector size, because the buffer length is now prefix_len + actual data len?

Reply:
good catch.